### PR TITLE
Add support for times to DateValue

### PR DESCRIPTION
**Pull request by gcampax**
And add a new TimeFn semantic function that complements DateFn

### CODE REVIEW

File: src/edu/stanford/nlp/sempre/DateValue.java [hunk 1]
```diff
@@ -2,33 +2,40 @@
...
   public static DateValue parseDateValue(String dateStr) {
-    if (dateStr.equals("PRESENT_REF")) return null;
+    if (dateStr.equals("PRESENT_REF")) return now();
     if (dateStr.startsWith("OFFSET")) return null;
 
     // We don't handle the following things:
     //   - "30 A.D" since its value is "+0030"
     //   - "Dec 20, 2009 10:04am" since its value is "2009-12-20T10:04"
     int year = -1, month = -1, day = -1;
+    int hour = 0, minute = 0;
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/DateValue.java]: For consistency, could we make the defaults -1 to signify "value doesn't exist"?  00:00:00 is actually a valid time.

File: src/edu/stanford/nlp/sempre/DateValue.java [hunk 2]
```diff
@@ -51,24 +66,56 @@ private static int parseIntRobust(String i) {
     return val;
   }
 
+  private static double parseDoubleRobust(String i) {
+    try {
+      return Double.parseDouble(i);
+    } catch (NumberFormatException ex) {
+      return 0;
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/DateValue.java]: In concordance with the above comment, return -1 to signify a bad value

File: src/edu/stanford/nlp/sempre/DateValue.java [hunk 3]
```diff
@@ -51,24 +66,56 @@ private static int parseIntRobust(String i) {
...
   public DateValue(LispTree tree) {
     this.year = Integer.valueOf(tree.child(1).value);
     this.month = Integer.valueOf(tree.child(2).value);
     this.day = Integer.valueOf(tree.child(3).value);
+    if (tree.children.size() > 3) {
+      this.hour = Integer.valueOf(tree.child(4).value);
+      this.minute = Integer.valueOf(tree.child(5).value);
+      this.second = Double.valueOf(tree.child(6).value);
+    } else {
+      this.hour = 0;
```
Comment (percyliang) [inline src/edu/stanford/nlp/sempre/DateValue.java]: -1 (actually, probably would be a good idea to define a constant INVALID = -1 and just use that everywhere)

**Comment by percyliang** [comment]
@gcampax: Thanks for the change - this was on the TODO list for a while.  Just a few minor comments...

**Comment by gcampax** [comment]
The choice of 0 0 0 for missing time was intentional actually, because the idea is that if you say just a date you mean midnight of that day.
It makes a lot of the code handling the resulting logical form easier because it does not need to special case -1.

**Comment by percyliang** [comment]
In the context of the semantic parser, when I say, "I want to leave on Jan
23", I definitely don't mean I want to leave on Jan 23 at midnight.  If we
represent the fact that the time is missing, then the system can naturally
ask follow up questions like "When do you want to leave?"  So I think
differentiating missing from default is important.

**Comment by percyliang** [comment]
@gcampax: What would you like to do with PR?

**Comment by gcampax** [comment]
I had not realized this is still open.
Meanwhile our codebase for SEMPRE has diverged significantly, and I recently realized in fact that changes to FloatingParser a few months back make merging impossible (or at least impractical). Given that, I think this PR can be closed.

**Pull request closed without merging**